File: hosts/python/src/florecon/projections.py
"""Explicit projections from the allocation-native Report hypergraph."""


def strict_assignments(report: dict) -> list[tuple[int, int]]:
    """Return one (row id, group id) per row only if the report is not split.

    Raises ValueError if a row id participates in multiple groups.
    """
    by_id = {}
    for a in report.get("allocations", []):
        by_id.setdefault(int(a["id"]), set()).add(int(a["group_id"]))
    out = []
    for id_, groups in sorted(by_id.items()):
        if len(groups) != 1:
            raise ValueError(f"row {id_} is split across groups {sorted(groups)}")
        out.append((id_, next(iter(groups))))
    return out


def primary_assignments(report: dict, policy: str = "largest_abs") -> list[tuple[int, int]]:
    """Return one (row id, group id) per row, choosing a primary group when a
    row is split across several (lot strategies). ``policy`` is one of
    ``largest_abs`` (default), ``prefer_clean``, or ``first_group``.

    Unlike :func:`strict_assignments` this never raises — use it for row-level
    exports where every row needs exactly one home.
    """
    by_id: dict[int, list[dict]] = {}
    for a in report.get("allocations", []):
        by_id.setdefault(int(a["id"]), []).append(a)
    groups = {int(g["group_id"]): g for g in report.get("groups", [])}

    def score(a: dict):
        g = groups.get(int(a["group_id"]), {})
        clean = 1 if abs(int(g.get("net", 0))) == 0 else 0
        if policy == "first_group":
            return (-int(a["group_id"]),)
        if policy == "prefer_clean":
            return (clean, abs(int(a["amount"])), -int(a["group_id"]))
        return (abs(int(a["amount"])), clean, -int(a["group_id"]))

    out = []
    for id_, allocs in sorted(by_id.items()):
        best = max(allocs, key=score)
        out.append((id_, int(best["group_id"])))
    return out


def connected_components(report: dict) -> list[dict]:
    """Connected components of the bipartite graph row id <-> group id."""
    row_to_groups = {}
    group_to_rows = {}
    for a in report.get("allocations", []):
        r = int(a["id"])
        g = int(a["group_id"])
        row_to_groups.setdefault(r, []).append(g)
        group_to_rows.setdefault(g, []).append(r)

    seen_rows = set()
    seen_groups = set()
    out = []
    for start in list(row_to_groups):
        if start in seen_rows:
            continue
        rows = set()
        groups = set()
        row_stack = [start]
        group_stack = []
        while row_stack or group_stack:
            while row_stack:
                r = row_stack.pop()
                if r in seen_rows:
                    continue
                seen_rows.add(r)
                rows.add(r)
                group_stack.extend(row_to_groups.get(r, []))
            while group_stack:
                g = group_stack.pop()
                if g in seen_groups:
                    continue
                seen_groups.add(g)
                groups.add(g)
                row_stack.extend(group_to_rows.get(g, []))
        out.append({"rows": sorted(rows), "groups": sorted(groups)})
    out.sort(key=lambda c: (c["rows"][0] if c["rows"] else 0, c["groups"][0] if c["groups"] else 0))
    return out

File: hosts/js/core/projections.js
// Explicit projections over the allocation-native Report hypergraph — the JS
// mirror of projections.py. No Arrow/DOM deps, so it is unit-testable under node.

export function strictAssignments(report) {
  const byId = new Map();
  for (const a of report.allocations || []) {
    if (!byId.has(a.id)) byId.set(a.id, new Set());
    byId.get(a.id).add(a.group_id);
  }
  const out = [];
  for (const [id, gs] of byId) {
    if (gs.size !== 1) throw new Error(`row ${id} is split across groups ${[...gs]}`);
    out.push([id, [...gs][0]]);
  }
  return out.sort((a, b) => a[0] - b[0]);
}

export function primaryAssignments(report, policy = "largest_abs") {
  const byId = new Map();
  for (const a of report.allocations || []) {
    if (!byId.has(a.id)) byId.set(a.id, []);
    byId.get(a.id).push(a);
  }
  const groups = new Map((report.groups || []).map((g) => [g.group_id, g]));
  const score = (a) => {
    const g = groups.get(a.group_id) || {};
    const clean = Math.abs(g.net || 0) === 0 ? 1 : 0;
    if (policy === "first_group") return [-a.group_id];
    if (policy === "prefer_clean") return [clean, Math.abs(a.amount || 0), -a.group_id];
    return [Math.abs(a.amount || 0), clean, -a.group_id];
  };
  const better = (a, b) => {
    const sa = score(a),
      sb = score(b);
    for (let i = 0; i < Math.max(sa.length, sb.length); i++) {
      const da = sa[i] ?? 0,
        db = sb[i] ?? 0;
      if (da !== db) return da > db;
    }
    return false;
  };
  const out = [];
  for (const [id, allocs] of byId) {
    let best = allocs[0];
    for (const a of allocs) if (better(a, best)) best = a;
    out.push([id, best.group_id]);
  }
  return out.sort((a, b) => a[0] - b[0]);
}

export function connectedComponents(report) {
  const r2g = new Map(),
    g2r = new Map();
  const push = (m, k, v) => (m.has(k) ? m.get(k) : m.set(k, []).get(k)).push(v);
  for (const a of report.allocations || []) {
    push(r2g, a.id, a.group_id);
    push(g2r, a.group_id, a.id);
  }
  const seenR = new Set(),
    seenG = new Set(),
    out = [];
  for (const start of r2g.keys()) {
    if (seenR.has(start)) continue;
    const rows = new Set(),
      groups = new Set(),
      rs = [start],
      gs = [];
    while (rs.length || gs.length) {
      while (rs.length) {
        const r = rs.pop();
        if (seenR.has(r)) continue;
        seenR.add(r);
        rows.add(r);
        gs.push(...(r2g.get(r) || []));
      }
      while (gs.length) {
        const g = gs.pop();
        if (seenG.has(g)) continue;
        seenG.add(g);
        groups.add(g);
        rs.push(...(g2r.get(g) || []));
      }
    }
    out.push({ rows: [...rows].sort((a, b) => a - b), groups: [...groups].sort((a, b) => a - b) });
  }
  return out;
}
